### PR TITLE
Add support for DevWorkspaces with parents

### PR DESCRIPTION
**Pull request by amisevsk**
### What does this PR do?
This PR is a copy of https://github.com/devfile/library/pull/74. It adds support for parents in DevWorkspaces (since most of the functionality was already there, and a majority of the work was renaming things to make sense). It also allows plugins/parents specified by URI to point at devfiles, devworkspaces, or devworkspacetemplates, and processes each accordingly. 

### What issues does this PR fix or reference?
No issue, but the work was done for https://github.com/devfile/devworkspace-operator/issues/334

### Is it tested? How?
I haven't prepared a test that uses parents, but regular functionality we currently use should still be supported.

### CODE REVIEW

File: pkg/library/flatten/testdata/general/fail-nicely-when-no-http-client-provided_id.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+input:
+  workspace:
+    components:
+      - name: test-plugin
+        plugin:
+          id: test/plugin
+
+output:
+  errRegexp: "cannot resolve resources by id: no HTTP client provided"
```
Comment (sleshchenko) [inline pkg/library/flatten/testdata/general/fail-nicely-when-no-http-client-provided_id.yaml:11]:
```suggestion
  errRegexp: "cannot resolve resources by id: no HTTP client provided"

```

File: pkg/library/flatten/testdata/general/fail-nicely-when-no-http-client-provided_uri.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+input:
+  workspace:
+    components:
+      - name: test-plugin
+        plugin:
+          uri: https://test-repo.io/my-plugino
+
+output:
+  errRegexp: "cannot resolve resources by URI: no HTTP client provided"
```
Comment (sleshchenko) [inline pkg/library/flatten/testdata/general/fail-nicely-when-no-http-client-provided_uri.yaml:11]:
```suggestion
  errRegexp: "cannot resolve resources by URI: no HTTP client provided"

```

File: pkg/library/flatten/testdata/general/fail-nicely-when-no-k8s-client-provided.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  workspace:
+    components:
+      - name: test-plugin
+        plugin:
+          kubernetes:
+            name: test-plugin-a
+            namespace: test-ns
+
+output:
+  errRegexp: "cannot resolve resources by kubernetes reference: no kubernetes client provided"
```
Comment (sleshchenko) [inline pkg/library/flatten/testdata/general/fail-nicely-when-no-k8s-client-provided.yaml:13]:
```suggestion
  errRegexp: "cannot resolve resources by kubernetes reference: no kubernetes client provided"

```

File: pkg/library/flatten/testdata/general/fail-nicely-when-no-namespace.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+input:
+  workspace:
+    components:
+      - name: test-plugin
+        plugin:
+          kubernetes:
+            name: test-plugin-a
+
+output:
+  errRegexp: "specifies a kubernetes reference without namespace and a default is not provided"
```
Comment (sleshchenko) [inline pkg/library/flatten/testdata/general/fail-nicely-when-no-namespace.yaml:12]:
```suggestion
  errRegexp: "specifies a kubernetes reference without namespace and a default is not provided"

```

File: pkg/library/flatten/testdata/general/fail-nicely-when-no-registry-url.yaml [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+input:
+  workspace:
+    components:
+      - name: test-plugin
+        plugin:
+          id: test/plugin
+
+output:
+  errRegexp: "plugin test-plugin does not specify a registryUrl and no internal registry is configured"
```
Comment (sleshchenko) [inline pkg/library/flatten/testdata/general/fail-nicely-when-no-registry-url.yaml:11]:
```suggestion
  errRegexp: "plugin test-plugin does not specify a registryUrl and no internal registry is configured"

```

**Review by sleshchenko — APPROVED** [review]
LGTM
I looked through the PR diff and only managed to find end-find empty lines )

I did not tested but we have pretty good unit testing )

File: pkg/library/flatten/testdata/general/fail-nicely-when-no-http-client-provided_id.yaml (hunk 1 above)
Comment (amisevsk) [reply pkg/library/flatten/testdata/general/fail-nicely-when-no-http-client-provided_id.yaml:11]: I was so sure I'd found them all :disappointed: 

Thanks

**Review by JPinkney — APPROVED** [review]
LGTM, other than Serhii's comments

**Comment by amisevsk** [comment]
:roll_eyes: even writing instructions in my commit message was not enough to get me to remember: https://github.com/devfile/devworkspace-operator/pull/346/commits/173f24212f9ff81d81a840fd88d535e82b18fec8.

Sorry all, I should've thought before merging.

**Comment by amisevsk** [comment]
> Why you just don't use git commit --fixup COMMIT ?

Today I learned...

**Pull request merged**